src/uct/ib/mlx5/gga/gga_mlx5.c

```diff
@@ -14,14 +14,50 @@
 
 #include <uct/ib/mlx5/rc/rc_mlx5.inl>
 
+#include <ucs/datastruct/khash.h>
+
+#include <pthread.h>
+
 #define UCT_GGA_MLX5_OPAQUE_BUF_LEN 64
 #define UCT_GGA_MAX_MSG_SIZE        (2u * UCS_MBYTE)
 
+
+/**
+ * Hashed per MD part of rkey
+ */
 typedef struct {
+    uct_ib_mlx5_devx_mem_t *memh;
+    uct_rkey_bundle_t      rkey_ob;
+} uct_gga_mlx5_md_rkey_handle_t;
+
+/**
+ * Pointer to @ref uct_gga_mlx5_rkey_handle_t is a key in the hash
+ */
+typedef uintptr_t uct_gga_mlx5_rkey_hash_key_t;
+
+KHASH_MAP_INIT_INT64(resolved_rkeys, uct_gga_mlx5_md_rkey_handle_t*);
+
+typedef struct {
+    uct_ib_mlx5_md_t        super;
+    khash_t(resolved_rkeys) rkey_cache;
+} uct_gga_mlx5_md_t;
+
+/**
+ * Entry of uct_gga_mlx5_rkey_handle_t::mds to implement many-to-many
+ * relationship between rkeys and mds
+ */
+typedef struct {
+    ucs_list_link_t   list_link;
+    uct_gga_mlx5_md_t *md;
+} uct_gga_mlx5_rkey_handle_md_link_t;
```

Comment:
as discussed offline, change the structs to be global hash table and not per md, and review locking scheme when accessing from multiple threads (md)